### PR TITLE
fix data races

### PR DESCRIPTION
**Pull request by pytlesk4**
set finalizers properly, and unset them too for when release is called manually.

### CODE REVIEW

File: v8.go [hunk 1]
```diff
@@ -77,14 +77,19 @@ type Snapshot struct{ data C.StartupData }
 
 func newSnapshot(data C.StartupData) *Snapshot {
 	s := &Snapshot{data}
-	runtime.SetFinalizer(s, (*Snapshot).release)
```
Comment (augustoroman) [inline v8.go]: Why this change?  Is it necessary?  I believe the original method should work.

File: v8.go [hunk 2]
```diff
@@ -184,9 +200,11 @@ type Context struct {
 	ptr C.ContextPtr
 
 	callbacks      map[int]callbackInfo
+	callbacksMutex sync.RWMutex
```
Comment (augustoroman) [inline v8.go]: A Context isn't safe for concurrent usage, so I don't think we need these mutexes -- you shouldn't be accessing them from more than one goroutine at all.

File: v8.go (hunk 1 above)
Comment (pytlesk4) [inline v8.go]: Sorry, I didn't know about [method expressions](https://golang.org/ref/spec#Method_expressions). I can revert that, the main reason I did that is because I wanted to put some debug statements to print information to console to verify that the finalizers were getting called. I am not sure if they are getting called or not.

For example: 

```go
func main() {
  ctx := NewIsolate().NewContext()
  // Doing something...
  
  // Done with ctx, set to nil and it should trigger the finalizer to run. Also, ctx holds a pointer to the isolate and the isolate doesn't get cleaned up either.
  ctx = nil
}
```

Am I understanding this correctly? There are couple of reasons why it is happening, and will detail it once I do a little more digging.

File: v8.go (hunk 2 above)
Comment (pytlesk4) [inline v8.go]: I did this a while ago, I can't remember at the time why, the way I am using this in my projects it shouldn't be accessed by another goroutine. I will try to reproduce the issue I was running into and create a test for it if I can.

**Comment by pytlesk4** [comment]
Building this branch of v8 - [6.0-lkgr](https://github.com/v8/v8/tree/6.0-lkgr)

File: v8.go (hunk 1 above)
Comment (augustoroman) [inline v8.go]: The finalizer will run when the context (and the isolate) are garbage collected.  That doesn't necessarily happen when `ctx=nil` is run.  In fact, the go runtime doesn't guarantee that finalizers will ever run -- it may be that the runtime decides that it doesn't need to GC anything.  You can force the GC to run using runtime.GC(), but I don't know whether that forces finalizers to run.  It's worth testing.

Comment (pytlesk4) [inline v8.go]: Reverted some of my previous changes, I will try to reproduce the race condition I was running into, I very well could have been using Context incorrectly.

I added a couple test cases around finalizers running, and it doesn't look like ctx is ever getting cleaned up. The Isolate is getting cleaned up, but only if there are no Contexts that were created from that Isolate, I just added a test for an isolate that creates a context to illustrate this. I can add some around values, but ultimately those won't get cleaned up either because the context is holding a pointer to the value(same thing is happening with isolates). 

The reason context is getting cleaned up is because of the [map](https://github.com/augustoroman/v8/blob/master/v8.go#L388) that stores the ctx so we can look up a golang function when it is called from javascript.

Also, from the limited research I have done, and based on your comment, it doesn't sound like a good idea to depend on the finalizers. 

What do you think about exposing the release functions? We could also remove the map of contexts, but I don't know if it is possible because of CGO limitations. This library looks like they don't use a map to keep track of the context: https://github.com/kingland/go-v8, but the last commit was in 2015 and CGO wasn't as strict back then.

Comment (augustoroman) [inline v8.go]: Thanks for the new tests.

I went through this carefully last night.  We'll have to expose the release functions.  I was really hoping to avoid that -- I actually had explicit releasing in the code before I refactored and published the first version.  It makes tests much uglier because you need to release the Isolate and the Context for each test.  In my experience, most usage of the contexts and isolates are long-lived (e.g. as pools like you mentioned), so figuring out when to release the resources can be annoying.  I'd really like for the context to be able to clean itself up, but I don't see a clean way of doing that with the multiple reference cycles that we have between the map, the Values, and the Context itself.  Darn!

As for depending on the finalizers, I've gone back and forth between them.  I was convinced that it makes sense to use them to tie C memory to Go objects from this thread: https://groups.google.com/d/msg/golang-dev/DMiUkpS1uyQ/0sHpTacG00YJ (in particular the post from rsc about whether they will actually run).

File: v8_c_bridge.cc [hunk 3]
```diff
@@ -25,21 +25,22 @@
...
-void* ArrayBufferAllocator::Allocate(size_t length) {
-  void* data = AllocateUninitialized(length);
-  return data == nullptr ? data : memset(data, 0, length);
-}
-void* ArrayBufferAllocator::AllocateUninitialized(size_t length) { return malloc(length); }
-void ArrayBufferAllocator::Free(void* data, size_t) { free(data); }
-
-// We only need one, it's stateless.
-ArrayBufferAllocator allocator;
+// TODO: Figure out why this is causing a linker error.
```
Comment (augustoroman) [inline v8_c_bridge.cc]: yeah, I can't figure out the linker error either, but they did change the ArrayBuffer::Allocator API in 6.0 according to https://docs.google.com/document/d/1g8JFi8T_oAE_7uAri7Njtig7fKaPDfotU6huOa1alds/edit

Comment (augustoroman) [inline v8_c_bridge.cc]: so, just go ahead and delete the dead code here, and go with using the default allocator.

File: v8.go (hunk 1 above)
Comment (augustoroman) [inline v8.go]: hrm... actually, all we need to wrap the context one level and put the finalizer on the wrapper.

The map and the values can all point back to the internal wrapper object, and the outer wrapper is only for the external API.  When all references to the external object wrapper are gone, then we can finalize the context.

On the other hand... maybe somebody wants to fire off a long-running operation?  Perhaps the explicit release is better anyways.

File: v8_c_bridge.cc (hunk 3 above)
Comment (pytlesk4) [inline v8_c_bridge.cc]: Will do.

File: v8.go (hunk 1 above)
Comment (pytlesk4) [inline v8.go]: By wrapping, you mean creating a new struct that is just wrapping the context? If so, I like that. It would still be nice to expose release, while still leaving the finalizers. With release being exposed then the user can decide when to clean up, but if they forget then the finalizer will protect them.

Comment (augustoroman) [inline v8.go]: Yep.  Move the existing Context content to an internal contextData struct and use that for all of the self-referential stuff.  Then create a new Context struct that just holds a *contextData.  Same with Isolate.  Values should point to contextData, so we should have a strict ordering of:

`Isolate -> Context[s] -> Value[s]`

Comment (pytlesk4) [inline v8.go]: Alright, I will try to get to this over the weekend.

File: v8_c_bridge.cc (hunk 3 above)
Comment (pytlesk4) [inline v8_c_bridge.cc]: Just got back from vacation, going to try and finish this pull request up once I get caught up on everything.

**Comment by augustoroman** [comment]
I went through and finally figured out a clean solution to getting finalizers to run.  The problem was the context map that held a pointer to a Context so that callbacks could know which Context they are part of.  I fixed that so they only create an entry when while V8 code is evaluating, and then remove the entry afterwards.  That should fix the finalizer problem.

As for the data race, as far as I can tell there isn't one.  Please reopen if this is still causing trouble.
